### PR TITLE
Wire the DSA indexer into the training loop

### PR DESCRIPTION
**Pull request by alokpathy**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
This is the fourth of four stacked PRs that introduces a simplified version of DSA. These stacked PRs split PR #[6869](https://github.com/NVIDIA/Megatron-LM/pull/6869).

This PR wires the simplified DSA indexer into the main training loop. There are four main features included:

**Freezing:** `dsa_train_indexer_only` and `dsa_train_main_only` freeze either the non-indexer model parameters or the indexer model parameters. This is necessary for each stage of long-context training with DSA.

**Resetting.** A checkpoint trained without an indexer, or with a different one, needs the indexer reinitialized while the backbone continues from where it stopped. Two initializations are available. Random draws from a seed derived to be tensor-parallel-invariant and data-parallel-aware, so every TP rank of a shard agrees and different DP replicas do not diverge; the reset then broadcasts indexer parameters only. The alternative seeds the indexer from the main Q projection, gathering the fused QKV across TP, with optional rescaling.

**Optimizer state after a reset.**  Clears the indexer's moments *and* resets only its parameter groups' clocks, leaving the backbone's moments and step count untouched. This resets just the indexer rather than the whole run.

**Indexer logging.** Indexer and backbone gradients differ enough in scale that one combined grad-norm hides what either is doing. `calc_dsa_split_grad_norms` and `calc_dsa_split_grad_num_zeros` report them separately, and the LR logging distinguishes the two parameter groups.

| File | Change |
|---|---|
| `megatron/training/training.py` | Parameter freezing for `--dsa-train-indexer-only` / `--dsa-train-main-only`, indexer reset-on-load, indexer LR warmup, and split grad-norm/zeros logging. Parameter freezing is necessary for multistage long-context extension. |
| `megatron/training/checkpointing.py` | Restores model-defining DSA args; asserts trainability mode is unchanged when loading optimizer state |
| `megatron/training/arguments.py` | 32 `--dsa-*` flags and their validation. |
| `megatron/training/models/dist_utils.py` | Calls `apply_dsa_param_freezing` before DDP wrapping |
| `megatron/training/utils/common_utils.py` | Split indexer/backbone gradient norms and zero counts for TensorBoard/wandb. |
| `megatron/training/utils/__init__.py` | Re-exports those helpers. |

## Part of a series

This PR is 4 of 4 splitting a single 13k-line PR #6869 

| PR | Contents |
|----|----------|
| #7152     | min-memory DSA kernels (torch + Triton) |
|  #7156   | DSA-over-GQA attention layer            |
|  #7158 | per-parameter-group optimizer clocks    |
| **this PR**   | indexer freezing, reset, checkpointing  |

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/utils/common_utils.py [hunk 1]
```diff
@@ -369,6 +370,175 @@ def reduce_max_stat_across_model_parallel_group(
...
+                param_map[param] = main_param
+
+    for group in getattr(optimizer, "fp32_from_fp32_groups", []):
+        for param in group:
+            param_map[param] = param
+
+    return param_map
+
+
+def get_model_to_optimizer_param_map(optimizer) -> dict:
```
Comment (maanug-nv) [inline megatron/training/utils/common_utils.py:530]: why does `_get_model_to_optimizer_param_map` even need to be private then?

File: megatron/training/models/dist_utils.py [hunk 2]
```diff
@@ -112,6 +112,17 @@ def unimodal_build_distributed_models(
         else:
             logger.warning("Final pre wrap hook returned None, skipping pre wrap hooks.")
 
+    # DSA parameter freezing (--dsa-train-indexer-only / --dsa-train-main-only).
```
Comment (maanug-nv) [inline megatron/training/models/dist_utils.py:115]: I think this should be done via pre-wrap hook instead of adding new code to this function.

File: megatron/training/arguments.py [hunk 3]
```diff
@@ -2489,6 +2610,12 @@ def _add_network_size_args(parser):
         "use_te_rng_tracker",
         "log_max_attention_logit",
         "barrier_with_L1_time",
+        "rope_type",
```
Comment (maanug-nv) [inline megatron/training/arguments.py:2613]: why are these being skipped?

**Comment by maanug-nv** [comment]
### Code review

Found 4 issues:

1. `validate_args()` accesses `args.use_legacy_models`, but the standard parser no longer defines that argument. Any launch with `--overlap-param-gather` reaches this line and raises `AttributeError` before validation finishes. (historical git context)

https://github.com/NVIDIA/Megatron-LM/blob/227a1de8ecd8deb7878ad8ae107afb7bb73f4f66/megatron/training/arguments.py#L1048-L1057

2. A dense checkpoint records `experimental_attention_variant=None`, while a DSA reset run sets it to `"dsa"`. This unconditional compatibility check asserts that the two values match before `_reset_dsa_indexer_after_load()` runs. As a result, `--dsa-reset-indexer-on-load` cannot perform the intended dense-to-DSA conversion. `--no-load-optim` does not skip this check. (bug)

https://github.com/NVIDIA/Megatron-LM/blob/227a1de8ecd8deb7878ad8ae107afb7bb73f4f66/megatron/training/checkpointing.py#L203-L218
https://github.com/NVIDIA/Megatron-LM/blob/227a1de8ecd8deb7878ad8ae107afb7bb73f4f66/megatron/training/training.py#L3569-L3576

3. `training_log()` already runs at line 5743, then runs again for the same iteration at line 5806. The first call updates and clears loss counters and resets interval timers. The second call double-counts the iteration and losses, then repeats collective and parameter-norm work using reset state. Should the second block be removed? (bug)

https://github.com/NVIDIA/Megatron-LM/blob/227a1de8ecd8deb7878ad8ae107afb7bb73f4f66/megatron/training/training.py#L5773-L5825

4. The first logging path calculates DSA-only split gradient statistics for every model. On a normal non-DSA run, the default gradient clipping produces `grad_norm` and the default logging interval is one, so every step scans all gradients and launches extra distributed reductions. I think this should be gated on the presence of a DSA indexer parameter group? (bug)

https://github.com/NVIDIA/Megatron-LM/blob/227a1de8ecd8deb7878ad8ae107afb7bb73f4f66/megatron/training/training.py#L5715-L5733

<sub>Generated by Codex.</sub>

File: megatron/training/training.py [hunk 4]
```diff
@@ -1388,10 +1465,16 @@ def _split_spec_part(part):
             swiglu=args.swiglu,
             use_gated_delta_product=_uses_gated_delta_product_spec(args),
```
Comment (Mellonta) [inline megatron/training/training.py:1466]: Also add `gdp_dsa_stack_spec` from previous PR.

File: megatron/training/training.py [hunk 5]
```diff
@@ -1460,11 +1545,507 @@ def _get_field(string, type):
...
-    print_rank_0(f"megatron.training.get_start_time_from_progress_log: "
-                 f"{start_time=}, {start_num_floating_point_operations=}")
+    print_rank_0(
+        f"megatron.training.get_start_time_from_progress_log: "
+        f"{start_time=}, {start_num_floating_point_operations=}"
+    )
     return datetime.strptime(start_time, '%Y-%m-%d %H:%M:%S'), start_num_floating_point_operations
 
 
+def _freeze_non_dsa_indexer_parameters(model):
```
Comment (Mellonta) [inline megatron/training/training.py:1555]: We need to freeze the router biases too. They’re buffers updated outside the optimizer, so freezing parameter gradients still lets them change during indexer-only training. like:
```
for module in model_module.modules():
    if hasattr(module, 'frozen_expert_bias'):
        module.frozen_expert_bias = True
```

File: megatron/training/training.py [hunk 6]
```diff
@@ -1460,11 +1545,507 @@ def _get_field(string, type):
...
+
+    The initialization derives the indexer from the loaded checkpoint, so it can only run
+    against the checkpoint the run starts from. ``args.iteration`` is zero exactly then:
+    ``load_checkpoint`` forces it to zero under ``--finetune`` and ``--pretrained-checkpoint``,
+    and reports the stored iteration on an ordinary resume. Re-running the initialization on
+    a resume would discard every indexer update the run has made.
+    """
+    if not getattr(args, "dsa_reset_indexer_on_load", False):
+        return False
+    return getattr(args, "iteration", 0) == 0
```
Comment (Mellonta) [inline megatron/training/training.py:1975]: Let's not enforce that, so people can choose if they want to use finetune.

File: megatron/training/training.py [hunk 7]
```diff
@@ -3661,16 +4445,14 @@ def training_log(
...
+                pg_collection.has_language_model()
+            ), "DSA indexer logging requires a language-model ProcessGroupCollection"
             pg_collection = pg_collection.get_language_model_collection()
         if pg_collection is None:
             # Compatibility path for legacy training entrypoints such as tasks/finetune_utils.py.
             # The core logger still receives explicit groups and does not read MPU globals.
-            pg_collection = ProcessGroupCollection.use_mpu_process_groups(
-                required_pgs=['pp', 'dp']
-            )
+            pg_collection = ProcessGroupCollection.use_mpu_process_groups(required_pgs=['pp', 'dp'])
```
Comment (Mellonta) [inline megatron/training/training.py:4455]: also include `gtp_remat` here

File: megatron/training/utils/common_utils.py [hunk 8]
```diff
@@ -369,6 +370,171 @@ def reduce_max_stat_across_model_parallel_group(
...
+                continue
+            seen_param_ids.add(param_id)
+
+            optim_param = param_to_optim_param.get(param)
+            if optim_param is None:
+                continue
+
+            if not param_is_not_shared(param):
+                continue
+            if not param_is_not_tensor_parallel_duplicate(param, tp_group=tp_group):
```
Comment (Mellonta) [inline megatron/training/utils/common_utils.py:401]: also filter `param_is_not_gtp_duplicate(param)`

**Review by Mellonta — COMMENTED** [review]
Can we leave main-q-mean and main-q-mean-rescaled out of this PR for now? We can add these back later if needed, with proper GTP/MXFP8 support.
